### DISCUSSION TITLE
Clientside serialization: transfer data structure packed as xstring

### DISCUSSION THREAD

**Discussion by josci** [SAP/PyRFC#262] [Q&A] [answered]
Hi,

for my current szenario I have to call a function module which requires a xstring input parameter (for Example RSDSO_WRITE_API_RFC) to transfer a large volume of test data. This xstring contains a table row stored as a byte string. 
An alternative approach to serialization data in a JSON / XML format leads to a big performance problem. 

Example:

**ABAP function  module**

```
  types: begin of ty_struct,
           text type char5,
           int  type i,
         end of ty_struct.

  data(target_struct) = value ty_struct( ).

  data(o_conv) = cl_abap_conv_in_ce=>create( encoding = 'UTF-8' endian = 'L' ).
  data(o_view) = cl_abap_view_offlen=>create_legacy_view( target_struct ).

  o_conv->convert_struc( exporting
                           input = pi_xstring
                           view  = o_view
                         importing
                           data  = target_struct ).

do something........`
```

**Python client**

```
byte_object = b'abc  \x00\x00\x00\x01\x00\x00\x00'

with Connection(**sap_system) as conn:
    conn.func_desc_remove('<SID>', 'FUNCTION_MODULE_NAME')
    result = conn.call('FUNCTION_MODULE_NAME', PI_XSTRING=byte_object)
```
The example byte string is deserialized to target_struct-text = 'abs', target_struct-int = 1

Currently I'm stuck with building the byte string on the python side. Are there any documents / examples availible how to build the byte string for a known ABAP target structure?

Any information is greatly appreciated!

Best regards
Johannes

**Comment by bsrdjan** [comment, marked as answer]
ABAP byte string counterpart on Python side are `bytes`. The only example is in unit test: https://github.com/SAP/PyRFC/blob/main/tests/test_datatypes.py#L36
